arch/sparc/core/interrupt_trap.S

```diff
@@ -0,0 +1,238 @@
+	/* %g2, %g3 are used at manual window overflow so save temporarily */
+	mov	%g2, %l4
+	mov	%g3, %l5
+
+	/* We may have trapped into the invalid window. If so, make it valid. */
+	rd	%wim, %g2
+	srl	%g2, %l0, %g3
+	cmp	%g3, 1
+	bne	.Lwodone
+	 nop
+
+	/* Do the window overflow. */
+	sll	%g2, (CONFIG_SPARC_NWIN-1), %g3
+	srl	%g2, 1, %g2
+	or	%g2, %g3, %g2
+
+	/* Enter window to save. */
+	save
+	/* Install new wim calculated above. */
+	mov	%g2, %wim
+	nop
+	nop
+	nop
+	/* Put registers on the dedicated save area of the ABI stack frame. */
+	std	%l0, [%sp + 0x00]
+	std	%l2, [%sp + 0x08]
+	std	%l4, [%sp + 0x10]
+	std	%l6, [%sp + 0x18]
+	std	%i0, [%sp + 0x20]
+	std	%i2, [%sp + 0x28]
+	std	%i4, [%sp + 0x30]
+	std	%i6, [%sp + 0x38]
+	/* Leave saved window. */
+	restore
+
+.Lwodone:
+	/*
+	 * %l4: %g2 at trap time
+	 * %l5: %g3 at trap time
+	 *
+	 * Save the state of the interrupted task including global registers on
+	 * the task stack.
+	 *
+	 * IMPORTANT: Globals are saved here as well on the task stack, since a
+	 * context switch might happen before the context of this interrupted
+	 * task is restored.
+	 */
+
+	/* Allocate stack for isr context including ABI frame. */
+	sub	%fp, ISF_SIZE, %sp
+	/*
+	 * %fp: %sp of interrupted task
+	 * %sp: %sp of interrupted task - ISF_SIZE.
+	 *      (fits a full ABI frame + what we store here)
+	 *
+	 * Save the interrupted context.
+	 */
+	std	%l0, [%sp + ISF_PSR_OFFSET]     /* psr pc */
+	st	%l2, [%sp + ISF_NPC_OFFSET]     /* npc */
+	st	%g1, [%sp + ISF_G1_OFFSET]      /* g1 */
+	std	%l4, [%sp + ISF_G2_OFFSET]      /* g2  g3 */
+	st	%g4, [%sp + ISF_G4_OFFSET]      /* g4 */
+	std	%i0, [%sp + ISF_I0_OFFSET]      /* i0  i1 */
+	std	%i2, [%sp + ISF_I2_OFFSET]      /* i2  i3 */
+	std	%i4, [%sp + ISF_I4_OFFSET]      /* i4  i5 */
+	std	%i6, [%sp + ISF_I6_OFFSET]      /* fp  i7 */
+	rd	%y, %g1
+	st	%g1, [%sp + ISF_Y_OFFSET]	/* y */
+
+	/* %l5: reference to _kernel */
+	set	_kernel, %l5
+	/* Switch to interrupt stack. */
+	mov	%sp, %fp
+	ld	[%l5 + _kernel_offset_to_irq_stack], %sp
+
+	/* Allocate a full C stack frame */
+	sub	%sp, STACK_FRAME_SIZE, %sp
+	/*
+	 * %fp: %sp of interrupted task - ISF_SIZE.
+	 * %sp: irq stack - 96. An ABI frame
+	 */
+
+	/* Enable traps, raise PIL to mask all maskable interrupts. */
+	or	%l0, PSR_PIL, %o0
+	wr	%o0, PSR_ET, %psr
+	nop
+	nop
+	nop
+
+	/* SPARC interrupt request level is the first agrument */
+	call	z_sparc_enter_irq
+	 mov	%l3, %o0
+
+	/*
+	 * Switch back to interrupted task stack
+	 * %fp: %sp of interrupted task - ISF_SIZE.
+	 * %sp: irq stack - 96. An ABI frame
+	 */
+	mov	%fp, %sp
+	/*
+	 * %fp: %sp of interrupted task - ISF_SIZE.
+	 * %sp: %sp of interrupted task - ISF_SIZE.
+	 */
+
+#ifdef CONFIG_PREEMPT_ENABLED
+	/* Switch context if not already on the current one. */
+	ld	[%l5 + _kernel_offset_to_current], %l6
+	ld	[%l5 + _kernel_offset_to_ready_q_cache], %l7
+	cmp	%l6, %l7
+	beq	.Lno_reschedule
+	 nop
+	/* key (PIL=15) is the first argument */
+	call arch_swap
```

Comment:
swap is the older, more complex context switch method:

```
/**
 * Cooperatively context switch
 *
 * Must be called with interrupts locked with the provided key.
 * This is the older-style context switching method, which is incompatible
 * with SMP. New arch ports, either SMP or UP, are encouraged to implement
 * arch_switch() instead.
 *
 * @param key Interrupt locking key
 * @return If woken from blocking on some kernel object, the result of that
 *         blocking operation.
 */
int arch_swap(unsigned int key);
```

any particular reason why you didn't implement `arch_switch()`?

Reply:
Thank you for the comments regarding `arch_swap()` and `arch_switch()`. I have updated the implementation so it now uses the the `arch_switch()` API instead.

Reply:
tyvm!